From c7753c7cf08732aaf3d73e05147ec0d3eb38d3e3 Mon Sep 17 00:00:00 2001
From: Naveen Kumar Gaddipati <naveenku@xilinx.com>
Date: Wed, 7 Apr 2021 12:53:38 +0530
Subject: [PATCH 09/11] dma: clean the BD's only when done bit is set

Instead of cleaning all the BD's clean the BD's for which
done bit is set.

Signed-off-by: Naveen Kumar Gaddipati <naveenku@xilinx.com>
Signed-off-by: Nagaradhesh Yeleswarapu <nagaradh@xilinx.com>
---
 drivers/dma/xilinx/xilinx_dma.c | 29 ++++++++++++++++++++++-------
 1 file changed, 22 insertions(+), 7 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_dma.c b/drivers/dma/xilinx/xilinx_dma.c
index b67d707e8775..ed5863950488 100644
--- a/drivers/dma/xilinx/xilinx_dma.c
+++ b/drivers/dma/xilinx/xilinx_dma.c
@@ -96,6 +96,7 @@
 #define XILINX_DMA_DMASR_HALTED		BIT(0)
 #define XILINX_DMA_DMASR_DELAY_MASK		GENMASK(31, 24)
 #define XILINX_DMA_DMASR_FRAME_COUNT_MASK	GENMASK(23, 16)
+#define XILINX_DMA_BD_CMPLT                    BIT(31)
 
 #define XILINX_DMA_REG_CURDESC			0x0008
 #define XILINX_DMA_REG_TAILDESC		0x0010
@@ -1662,6 +1663,8 @@ static void xilinx_dma_issue_pending(struct dma_chan *dchan)
 static void xilinx_dma_complete_descriptor(struct xilinx_dma_chan *chan)
 {
 	struct xilinx_dma_tx_descriptor *desc, *next;
+	struct xilinx_axidma_tx_segment *segment;
+	struct xilinx_axidma_desc_hw *hw;
 
 	/* This function was invoked with lock held */
 	if (list_empty(&chan->active_list))
@@ -1674,11 +1677,21 @@ static void xilinx_dma_complete_descriptor(struct xilinx_dma_chan *chan)
 		else
 			desc->residue = 0;
 		desc->err = chan->err;
-
-		list_del(&desc->node);
-		if (!desc->cyclic)
-			dma_cookie_complete(&desc->async_tx);
-		list_add_tail(&desc->node, &chan->done_list);
+		segment = list_last_entry(&desc->segments,
+				struct xilinx_axidma_tx_segment, node);
+		hw = &segment->hw;
+		/*
+		 * Move BD chain to done list only if all descriptors
+		 * in the chain are complete.
+		 */
+		if (hw->status & XILINX_DMA_BD_CMPLT) {
+			list_del(&desc->node);
+			if (!desc->cyclic)
+				dma_cookie_complete(&desc->async_tx);
+			list_add_tail(&desc->node, &chan->done_list);
+		} else {
+			break;
+		}
 	}
 }
 
@@ -1860,8 +1873,10 @@ static irqreturn_t xilinx_dma_irq_handler(int irq, void *data)
 	if (status & XILINX_DMA_DMASR_FRM_CNT_IRQ) {
 		spin_lock(&chan->lock);
 		xilinx_dma_complete_descriptor(chan);
-		chan->idle = true;
-		chan->start_transfer(chan);
+		if (list_empty(&chan->active_list)) {
+			chan->idle = true;
+			chan->start_transfer(chan);
+		}
 		spin_unlock(&chan->lock);
 	}
 
-- 
2.17.1

